compression_verifier.go

```diff
@@ -0,0 +1,224 @@
+	supportedAlgorithms map[string]struct{}
+
+	// tableCmpressions is represented as table[column][compression_algorithm]
+	tableCompressions map[string]map[string]string
+}
+
+// NewCompressionVerifier first checks the map for supported compression algorithms before
+// initializing and returning the initialized instance.
+func NewCompressionVerifier(tableCompressions map[string]map[string]string) (*CompressionVerifier, error) {
+	supportedAlgorithms := make(map[string]struct{})
+	supportedAlgorithms[CompressionSnappy] = struct{}{}
+
+	compressionVerifier := &CompressionVerifier{
+		logger:              logrus.WithField("tag", "compression_verifier"),
+		supportedAlgorithms: supportedAlgorithms,
+		tableCompressions:   tableCompressions,
+	}
+
+	if err := compressionVerifier.verifyConfiguredCompression(tableCompressions); err != nil {
+		return nil, err
+	}
+
+	return compressionVerifier, nil
+}
+
+func (c *CompressionVerifier) verifyConfiguredCompression(tableCompressions map[string]map[string]string) error {
+	for table, columns := range tableCompressions {
+		for column, algorithm := range columns {
+			// Ensure algorithm is uppercased
+			algorithm = strings.ToUpper(algorithm)
+			tableCompressions[table][column] = algorithm
+
+			if _, ok := c.supportedAlgorithms[algorithm]; !ok {
+				return &UnsupportedCompressionError{
+					table:     table,
+					column:    column,
+					algorithm: algorithm,
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// GetHashes compares the source data with the target data to ensure the integrity of the
+// data being copied.
+//
+// The CompressionVerifier method checks if the existing table contains compressed data
+// and will apply the decompression algorithm to the applicable columns if necessary.
+// After the columns are decompressed, the hashes of the data are used and verified
+// for comparison equality
+func (c *CompressionVerifier) GetHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64, tableCompression map[string]string) (map[uint64][]byte, error) {
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := c.getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	return c.decompressAndHash(table, columns, rows, tableCompression)
+}
+
+func (c *CompressionVerifier) getRows(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (*sql.Rows, error) {
+	quotedPK := quoteField(pkColumn)
+	sql, args, err := rowSelector(columns, pkColumn).
+		From(QuotedTableNameFromString(schema, table)).
+		Where(sq.Eq{quotedPK: pks}).
+		OrderBy(quotedPK).
+		ToSql()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// This query must be a prepared query. If it is not, querying will use
+	// MySQL's plain text interface, which will scan all values into []uint8
+	// if we give it []interface{}.
```

Comment:
Instead of (or in addition to) having this comment, can we have a test that fails if someone changes this from prepared to non-prepared? (i.e. a test that captures the bug that would appear if this were not a prepared statement)

Reply:
This was copied from code in the `IterativeVerifier` 🍝. I'll look into catching this in a test (good idea).

Reply:
Added https://github.com/Shopify/ghostferry/issues/52